### DISCUSSION TITLE
Specify defalt state

### DISCUSSION THREAD

**Discussion by SYkirk** [FLAMEGPU/FLAMEGPU2#906] [Q&A] [answered]
I've created two new states as the user guide shows, but it throws "InvalidAgentState" exception with the message **"does not contain state 'default**'" everytime I run it. I've gone through the user guide to search for information about agent state but failed to solve it . How should I deal with the situation when defining and initialising agents with more states? Many thanks!

**Comment by Robadob** [comment, marked as answer]
There are two possible causes I can think for this off the top of my head.

1) You're using an old build of FLAMEGPU2 and spatial messaging, there was a bug that causes this exception to be thrown and [has been resolved](https://github.com/FLAMEGPU/FLAMEGPU2/commit/507e9746aa3c0282b4269a44509b35e16be6c59a).
2) You're calling a function which has a default argument `default` and need to specify the desired state. The most obvious one that comes to mind is `FLAMEGPU->agent(agentName, stateName)` within the `HostAPI` (this is what caused the above bug).

If you could grab the call-stack or location the exception is being thrown from, it would help narrow down the cause.

*The problem was confirmed as issue (1) via email.*

**Reply by SYkirk** [reply]
Actually I can fix it by adding an extra state called "defalt". Now I have six states among which five are what I want and one named "default" is set when intializing agents. Every agent is born with the state "defalt" and can be changed later according to their variables by agent function. It indeed makes sense! Yet I'm wondering if it is the best way. Thank you!

**Reply by Robadob** [reply]
The state `default` should not be required. It's simply the case, that most models don't require states, so when no states are specified this state is assumed everywhere.

Are you passing the desired state name to [`CUDASimulation::setPopulationData(AgentVector& population, const std::string &state_name = ModelData::DEFAULT_STATE)`](https://github.com/FLAMEGPU/FLAMEGPU2/blob/dee1934ec4751e56f217b9d97b842899de4b1ea7/include/flamegpu/gpu/CUDASimulation.h#L157) ?

**Reply by SYkirk** [reply]
I guess no.

I just define states during defining agents by Agent.newState() 
`person.newState("healthy")`

without specifing any state for the agent while initializing in host function
`person = FLAMEGPU.agent("person")`
`new_person = person.newAgent()`

If a state named "default" has not been defined, exception above will be thrown.

**Reply by Robadob** [reply]
Yes that's your issue.

```python
person = FLAMEGPU.agent("person", "healthy")
new_person = person.newAgent()  # Create a new 'person' in the 'healthy' state
```

The call to `FLAMEGPU.agent("person")` has a default second argument `"default"`. The agents are stored per-state, so operations are per agent-state, rather than per-agent.

**Reply by Robadob** [reply]
More info here: https://docs.flamegpu.com/guide/host-functions/agent-operations.html

**Reply by SYkirk** [reply]
Actually I've tried below before I came here:

`person.newState("healthy")`

and 

`person = FLAMEGPU.agent("person", "healthy")`
`new_person = person.newAgent()`

But it still threw the exception.
![Snipaste_2022-08-20_11-11-34](https://user-images.githubusercontent.com/82354985/185726798-742493fe-019c-4655-b727-925f673dc036.jpg)

Any way, I will keep trying to debug. Thanks a lot!

**Reply by Robadob** [reply]
The second code example you have provided should work.

We have a [test-case doing the same thing](https://github.com/FLAMEGPU/FLAMEGPU2/blob/dee1934ec4751e56f217b9d97b842899de4b1ea7/tests/test_cases/runtime/test_host_agent_creation.cu#L226), which appears to have been present since the feature was added.

Furthermore, the exception you have posted is coming from [`flamegpu/runtime/HostApi.cu(50)`](https://github.com/FLAMEGPU/FLAMEGPU2/blob/4b62c71e266836d0b8c38ea7d1d85e3784af6d21/src/flamegpu/runtime/HostAPI.cu#L52).

That is caused by calling `FLAMEGPU.agent("person")` without specifying the second state argument, but would need a full call-stack/backtrace to know the reason that's being called (e.g. your code, or something internal).

We [fixed a bug](https://github.com/FLAMEGPU/FLAMEGPU2/commit/507e9746aa3c0282b4269a44509b35e16be6c59a), where this was being caused internally if FLAME GPU tried to sort your agents for spatial messaging. That hasn't yet made it into a binary release though.

If you're able to share your code with me, I can get you an answer.

**Reply by SYkirk** [reply]
I'd love to if it is okay with you.
Could you show me a private way to share my code? Thanks!

**Reply by SYkirk** [reply]
Hi! I am wondering if there is any way to visualize agents with different color according to their different variable value rather than their different states, as setting different states brings additional states change steps and makes my code indeed verbose. Thanks!

**Reply by Robadob** [reply]
Yes, you can control agent colour via a selected variable.

This should be covered in the user guide here
https://docs.flamegpu.com/guide/visualisation/visualising-agents.html#agent-color

Let me know if anything requires further clarification.

Thanks

*I previously sent this reply via email and it appeared as a separate thread, have moved it*

**Reply by SYkirk** [reply]
I used `uDiscreteColor` to construct my own color as the guide shows, like:

`cmap = pyflamegpu.uDiscreteColor("health", pyflamegpu.WHITE)`

`cmap[0] = pyflamegpu.Color("#97FCAB") `

But it throws the exception:

![image](https://user-images.githubusercontent.com/82354985/189316903-a507b0e9-680d-416e-96f9-799ecdee5016.png)

**Reply by Robadob** [reply]
This appears to be a bug due to an oversight. The `DiscreteColor` types extend `std::map`. So we had assumed the SWIGs stdlib wrapping would be automatic, but it appears not.

I'll let you know when it's been fixed, this will require you to get a new python wheel.

**Reply by Robadob** [reply]
I've now fixed it in the pull request [here](https://github.com/FLAMEGPU/FLAMEGPU2/pull/922).

I will merge it ASAP, and set new wheels building for you.

You should be able to download a new binary wheel in ~3 hours.

**Reply by Robadob** [reply]
You can find all the new wheels built with the bugfix here: https://github.com/FLAMEGPU/FLAMEGPU2/actions/runs/3021752398#artifacts

As before, they are only available for 5-7 days after which they will expire.

**Reply by SYkirk** [reply]
The new wheel has been installed and it works!

Thanks a lot!
